gcc/rust/backend/rust-compile-base.cc

```diff
@@ -342,7 +366,10 @@ HIRCompileBase::compile_function (
   // we don't mangle the main fn since we haven't implemented the main shim
   bool is_main_fn = fn_name.compare ("main") == 0;
   std::string asm_name = fn_name;
-  if (!is_main_fn)
+  // TODO(liushuyu): we should probably move this part to
+  // `setup_attributes_on_fndecl` if possible
+  bool should_mangle = should_mangle_item (outer_attrs);
+  if (!is_main_fn && should_mangle)
```

Comment:
I think your right, we could change this by renaming "setup_attributes_on_fndecl" to "setup_fndecl" which implies a more generic setup function.

Then you can change this to remove the asm_name it will default to the ir_symbol_name at this point anyway.
```
  tree fndecl = ctx->get_backend ()->function (compiled_fn_type, ir_symbol_name,
					       asm_name, flags, locus);
```

```
  tree fndecl = ctx->get_backend ()->function (compiled_fn_type, ir_symbol_name,
					       "" /* asm_name */, flags, locus);
```

Then you pull this highlighted code into where you handle no_mangle. You might need to have a flag to check for no_mangle by iterating the attributes then depending on the flag mangle or no mangle at the end.

Reply:
> I think your right, we could change this by renaming "setup_attributes_on_fndecl" to "setup_fndecl" which implies a more generic setup function.
> 
> Then you can change this to remove the asm_name it will default to the ir_symbol_name at this point anyway.
> 
> ```
>   tree fndecl = ctx->get_backend ()->function (compiled_fn_type, ir_symbol_name,
> 					       asm_name, flags, locus);
> ```
> 
> ```
>   tree fndecl = ctx->get_backend ()->function (compiled_fn_type, ir_symbol_name,
> 					       "" /* asm_name */, flags, locus);
> ```
> 
> Then you pull this highlighted code into where you handle no_mangle. You might need to have a flag to check for no_mangle by iterating the attributes then depending on the flag mangle or no mangle at the end.

I see. Thanks for the suggestion! I will attempt to implement the changes.

Reply:
Done. I have merged the logic into the renamed `setup_fndecl` function.